### PR TITLE
Minimize threadworker and use Context as indication of thread

### PR DESCRIPTION
**Pull request by cbaggers**
This PR has two main goals.

1. Make the language agnostic 'Scripting.Context' more abstract by moving any JS specifics to a language specific context
2. Increase the number of places where we use the pattern of taking the `Scripting.Context` as the primary argument if the method is required to be run on the VM thread
3. Refactor the JS ThreadWorker so that it doesn't implement so many things. The goal is that it will only manage the VM's thread.

The changes in this PR are:

- There is a new abstract `JSContext` class which the JS implementations subclass.
- `Context` & `ThreadWorker` are no longer partial classes
- `Fuse.Scripting`'s `Function` type has a `Call` method, this now takes a `Scripting.Context`. This guarentees that it can only occur on the VM thread.
- `Scripting.Context`'s `IThreadWorker` is now obtained via the abstract `ThreadWorker` property rather than having to be passed down via the constructor
- the JS specific `CreateContext` method has moved to the `JSContext` class
- IMirror is no longer implemented by ThreadWorker. This functionality has been moved to the context (however it is now easier to spin into it's own class)
- Moved `ArrayMirror`, `ClassInstance`, `ModuleInstance`, `ObjectMirror`, `Observable`, `ObservableProperty`, `RootableScriptModule` & `ThreadWorker` to the `Fuse.Scripting.JavaScript` namespace
- Where currently possible use `IThreadWorker` instead of `ThreadWorker` (even if for now it means a few more explicit casts)
- Removed the `CanEvaluate` method and instead rely on the passing of the `Scripting.Context` to know if we are on the VM thread or not. This required adding some methods to the `ISubscription` which take the `Scripting.Context`. We would like to remove the methods that do not take the context however it seems there are currently places where this is called from the UI thread (which is a concern)
- The 'wrapping' functionality has been moved from the `ThreadWorker` to a standalone static class called `TypeWrapper`. The `IThreadWorker` no longer provides `Wrap` & `UnWrap`
- `ThreadWorker.ScriptClass` functionality moved to context. We will likely want to factor this out to a helper class however for now the major benefit is that `ThreadWorker` no longer owns these features.
- Remove the public `Context` property from the `ThreadWorker`. Sadly the context is still available via the internal field so that the tests can work. This will need to be fixed.
- `IThreadWorker` now has a variant of Invoke that the allows passing of a single argument. It is hoped that this is a temporary measure and that we can unify the handling of this across `Scripting` & `Scripting.JavaScript` in a more satisfying way.
- `Fuse.Reactive` now depends on `Fuse.Scripting` so that it can talk about the `Scripting.Context` in it's provided interfaces.
- `DateTimeConverterHelpers` moved to its own uno file.

The ugly parts of this PR are:
- HTTP is having to cache the context to pass a valid context to the JS Callbacks now that `Call` requires a context
- some places (like GetResponseContentByteArray) still get passed `null` instead of a Context where is it not clear how to obtain the Context.
- Too many places are using fully qualified type names due to ambiguity at the point in history that they were written. Some of these could likely be cleaned up soon
- The static `ThreadWorker` property in the `JavaScript` class is still relied on in too many places. A future PR should remove this and pass the ThreadWorker more appropriately.
- Some tests currently require non-standard access to the Context and so currently `Various.Test.uno` is getting the context via an internal field. this is gross
- `JSContext` is getting a little heavy. This is an expected step, but I would like to see the ScriptClass logic move back out to a auxiliary class.
- TreeObservable has to cache the context as it uses it in a way that is unsafe. This unsafe usage is on master so this isnt a regression
- too many places are still caching the context. We need to fix this in future PRs

This PR contains:
- [x] Changelog
- [ ] Documentation
- [ ] Tests

### CODE REVIEW

**Comment by cbaggers** [comment]
@kusma appveyor has passed so I think this is fairly safe to review now

File: Source/Fuse.Scripting.JavaScript/V8/Object.uno [hunk 1]
```diff
@@ -82,7 +82,7 @@ namespace Fuse.Scripting.V8
 		public override bool InstanceOf(Scripting.Function type)
 		{
 			var f = type as Function;
-			return (bool)_context._instanceOf.Call(this, type);
+			return (bool)_context._instanceOf.Call(_context, this, type);
 		}
 
 		public override object CallMethod(string name, params object[] args)
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/V8/Object.uno]: True but the original goal for this PR is to get to Function.Call taking the context, it just grew into a monster. Next I will fix up CallMethod which should be much easier now

File: Source/Fuse.Scripting.JavaScript/V8/Function.uno [hunk 2]
```diff
@@ -29,7 +29,7 @@ namespace Fuse.Scripting.V8
 				_function.AsValue().Release(_context._context);
 		}
 
-		public override object Call(params object[] args)
+		public override object Call(Scripting.Context context, params object[] args)
 		{
 			var cxt = _context._context;
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/V8/Function.uno]: It wants a V8 context so I'll just cast the arg instead

File: Source/Fuse.Scripting.JavaScript/TreeObservable/TreeObservable.uno [hunk 3]
```diff
@@ -7,8 +7,12 @@ namespace Fuse.Scripting.JavaScript
 {
 	class TreeObservable: TreeObject, IMirror
 	{
-		public TreeObservable(Scripting.Object obj) : base(obj)
+		// TreeObservable requires unsafe access to the context. This is a bug and needs fixing
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/TreeObservable/TreeObservable.uno]: agreed

File: Source/Fuse.Scripting.JavaScript/ThreadWorker.uno [hunk 4]
```diff
@@ -27,19 +27,19 @@ namespace Fuse.Scripting.JavaScript
 		public void Push(Scripting.Array arr, object value)
 		{
 			if (_push == null) _push = (Function)_context.Evaluate("push", "(function(arr, value) { arr.push(value); })");
-			_push.Call(arr, value);
+			_push.Call(_context, arr, value);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/ThreadWorker.uno]: Added a warning comment to the code point at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/JavaScriptCore/JavaScriptCore.uno [hunk 5]
```diff
@@ -427,7 +427,7 @@ namespace Fuse.Scripting.JavaScriptCore
 			return ::JSClassCreate(&classDef);
 		@}
 
-		public delegate JSValueRef RawCallback(JSValueRef[] args, out JSValueRef exception);
+		public delegate JSValueRef RawCallback(JSValueRef[] args, out JSValueRef exception); // TODO: Should add `Context context,`
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/JavaScriptCore/JavaScriptCore.uno]: agreed

File: Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno [hunk 6]
```diff
@@ -242,46 +245,46 @@ namespace Fuse.Reactive.FuseJS
...
+					func.Call(_context);
 				DetachRequest();
 			}
 
 			void OnProgress(HttpMessageHandlerRequest res, int current, int total, bool hastotal)
 			{
 				var func = Obj["onprogress"] as Function;
 				if (func != null)
-					func.Call(current, total, hastotal);
+					func.Call(_context, current, total, hastotal);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno]: Added a warning comment to the definition of `_context` pointing at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno [hunk 7]
```diff
@@ -242,46 +245,46 @@ namespace Fuse.Reactive.FuseJS
...
-					func.Call((int)_req.State);
+					func.Call(_context, (int)_req.State);
 			}
 
 			void OnDone(HttpMessageHandlerRequest res)
 			{
 				var func = Obj["ondone"] as Function;
 				if (func != null)
-					func.Call();
+					func.Call(_context);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno]: Added a warning comment to the definition of `_context` pointing at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno [hunk 8]
```diff
@@ -242,46 +245,46 @@ namespace Fuse.Reactive.FuseJS
...
 				DetachRequest();
 			}
 
 			void OnStateChanged(HttpMessageHandlerRequest res)
 			{
 				var func = Obj["onstatechanged"] as Function;
 				if (func != null)
-					func.Call((int)_req.State);
+					func.Call(_context, (int)_req.State);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno]: Added a warning comment to the definition of `_context` pointing at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno [hunk 9]
```diff
@@ -242,46 +245,46 @@ namespace Fuse.Reactive.FuseJS
...
+					func.Call(_context, error);
 				DetachRequest();
 			}
 
 			void OnTimeout(HttpMessageHandlerRequest res)
 			{
 				var func = Obj["ontimeout"] as Function;
 				if (func != null)
-					func.Call();
+					func.Call(_context);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno]: Added a warning comment to the definition of `_context` pointing at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno [hunk 10]
```diff
@@ -242,46 +245,46 @@ namespace Fuse.Reactive.FuseJS
...
 				DetachRequest();
 			}
 
 			void OnError(HttpMessageHandlerRequest res, string error)
 			{
 				var func = Obj["onerror"] as Function;
 				if (func != null)
-					func.Call(error);
+					func.Call(_context, error);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno]: Added a warning comment to the definition of `_context` pointing at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno [hunk 11]
```diff
@@ -242,46 +245,46 @@ namespace Fuse.Reactive.FuseJS
 			{
 				var func = Obj["onabort"] as Function;
 				if (func != null)
-					func.Call();
+					func.Call(_context);
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/FuseJS/Http.uno]: Added a warning comment to the definition of `_context` pointing at a fuselibs issue

File: Source/Fuse.Scripting.JavaScript/Dates.uno [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+using Uno;
+using Uno.Collections;
+using Uno.Text;
+using Uno.Threading;
+using Uno.IO;
+using Uno.UX;
+
+namespace Fuse.Scripting.JavaScript
+{
+	static class DateTimeConverterHelpers
```
[the remark this answers is not in the document]
Comment (cbaggers) [reply Source/Fuse.Scripting.JavaScript/Dates.uno]: agreed, thanks

File: Source/Fuse.Scripting.JavaScript/V8/Object.uno (hunk 1 above)
[the remark this answers is not in the document]
Comment (kusma) [reply Source/Fuse.Scripting.JavaScript/V8/Object.uno]: Sure, that's fine :)

**Comment by BeeWarloc** [comment]
@kusma @cbaggers Just attempted compiling `StreamingBasics` with latest masters, however it turns out that the API change for `NativeEvent.RaiseAsync` from this PR breaks [`Fuse.StreamingPlayer`](https://github.com/fuse-compound/Fuse.StreamingPlayer). Is this a known thing?

```
C:\src\Fuse.StreamingPlayer\src\JS.uno(48.23): E2009: 'Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>' has no constructors matching the argument list (string,<method_group>,<method_group>,<method_group>)
Candidates are: Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>..ctor(string)
                Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>..ctor(string,Fuse.Scripting.Array)
                Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>..ctor(string,[Uno.Func<Uno.Collections.List<StreamingPlayer.Track>>],[Uno.Action<Fuse.Scripting.Context, Fuse.Scripting.Array>],[Fuse.Scripting.ValueConverter<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>])
C:\src\Fuse.StreamingPlayer\src\JS.uno(48,24): Error E2009: 'Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>' has no constructors matching the argument list (string,<method_group>,<method_group>,<method_group>)
Candidates are: Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>..ctor(string)
                Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>..ctor(string,Fuse.Scripting.Array)
                Fuse.Scripting.NativeProperty<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>..ctor(string,[Uno.Func<Uno.Collections.List<StreamingPlayer.Track>>],[Uno.Action<Fuse.Scripting.Context, Fuse.Scripting.Array>],[Fuse.Scripting.ValueConverter<Uno.Collections.List<StreamingPlayer.Track>, Fuse.Scripting.Array>])0.7 s

```

I see there are some changelog entries describing these changes, which is good, but it might be a good idea to also explicitly tell our users that these are breaking changes. :)

**Pull request merged**